### ISSUE TITLE
SDImageIOAnimatedCoder crash only happened on iOS 15

### ISSUE THREAD

**Issue by WJustin** [SDWebImage/SDWebImage#3273]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           |ios
 Platform Version        |15.0
 SDWebImage Version     |5.10.4
 Integration Method     |cocoapods
 Xcode Version         |Xcode 12.3
 Repro rate             |sometimes 0.5%

### Issue Description and Steps
0
CoreFoundation
CFDataGetBytes + 156
4
UIKitCore
0x0000000184200000 + 16248057818473156320
5
GoShare2
0x0000000104a34000 + 1599128611935478020
6
GoShare2
-[SDImageIOAnimatedCoder animatedImageFrameAtIndex:] (SDImageIOAnimatedCoder.m:0)
7
GoShare2
-[SDAnimatedImage animatedImageFrameAtIndex:] (SDAnimatedImage.m:0)
8
GoShare2
__43-[SDAnimatedImagePlayer displayDidRefresh:]_block_invoke (SDAnimatedImagePlayer.m:0)
9
Foundation
0x0000000183580000 + 333168
21
libsystem_pthread.dylib
_pthread_wqthread + 4047424897389953252

**Comment by jschedler-imvu** [comment]
This is also happening with SDWebImage v5.11.1

CoreFoundation
0x180a2a000 + 770340
ImageIO
0x182559000 + 9984
UIKitCore
0x182ed4000 + 2186356
UIKitCore
0x182ed4000 + 1691360
SDWebImage
+[SDImageIOAnimatedCoder createFrameAtIndex:source:scale:preserveAspectRatio:thumbnailSize:options:] SDImageIOAnimatedCoder.m:270
SDWebImage
-[SDImageIOAnimatedCoder animatedImageFrameAtIndex:] SDImageIOAnimatedCoder.m:653
SDWebImage
-[SDAnimatedImage animatedImageFrameAtIndex:] SDAnimatedImage.m:263
SDWebImage
-[SDAnimatedImagePlayer displayDidRefresh:] SDAnimatedImagePlayer.m:352
Foundation
0x182254000 + 333168
Foundation
0x182254000 + 407256
Foundation
0x182254000 + 246052
Foundation
0x182254000 + 315408
Foundation
0x182254000 + 329144
Foundation
0x182254000 + 386608
libdispatch.dylib
0x180732000 + 75264
libdispatch.dylib
0x180732000 + 14672
libdispatch.dylib
0x180732000 + 28164
libdispatch.dylib
0x180732000 + 25848
libdispatch.dylib
0x180732000 + 85832
libdispatch.dylib
0x180732000 + 87912
libsystem_pthread.dylib
0x1f05f8000 + 4468
libsystem_pthread.dylib

**Comment by Johnlee666** [comment]
I have the same problem In iOS15. And the type of crash is SIGTRAP or SIGSEGV. 
SIGTRAP：
0
CoreFoundation
CFDataGetBytes + 156
1
ImageIO
CGImageGetImageSource + 156
2
ImageIO
CGImageGetImageSource + 156
3
UIKitCore
0x0000000185fb8000 + 2186356
4
UIKitCore
0x0000000185fb8000 + 1691360
5
+[SDImageIOAnimatedCoder createFrameAtIndex:source:scale:preserveAspectRatio:thumbnailSize:options:] (SDImageIOAnimatedCoder.m:270)
6
-[SDImageIOAnimatedCoder animatedImageFrameAtIndex:] (SDImageIOAnimatedCoder.m:0)
7
-[SDAnimatedImage animatedImageFrameAtIndex:] (SDAnimatedImage.m:0)
8
__43-[SDAnimatedImagePlayer displayDidRefresh:]_block_invoke (SDAnimatedImagePlayer.m:0)

**Comment by Johnlee666** [comment]
SIGSEGV
SEGV_ACCERR：
0
libsystem_platform.dylib
_platform_memmove + 420
1
CoreFoundation
CFDataGetBytes + 440
2
ImageIO
CGImageGetImageSource + 156
3
UIKitCore
0x00000001858b3000 + 2182368
4
UIKitCore
0x00000001858b3000 + 1688280
5
+[SDImageIOAnimatedCoder createFrameAtIndex:source:scale:preserveAspectRatio:thumbnailSize:options:] (SDImageIOAnimatedCoder.m:270)
6
-[SDImageIOAnimatedCoder animatedImageFrameAtIndex:] (SDImageIOAnimatedCoder.m:0)
7
-[SDAnimatedImage animatedImageFrameAtIndex:] (SDAnimatedImage.m:0)
8
__43-[SDAnimatedImagePlayer displayDidRefresh:]_block_invoke (SDAnimatedImagePlayer.m:0)

**Comment by kinarobin** [comment]
@WJustin @jschedler-imvu @Johnlee666 Please provide a full crash log.

**Comment by WJustin** [comment]
#17 Thread

SIGSEGV

SEGV_ACCERR

0
libsystem_platform.dylib
_platform_memmove + 420
4
UIKitCore
0x00000001833f8000 + 17533706319471169248
5
GoShare2
0x0000000102adc000 + 11679445168071893252
6
GoShare2
-[SDImageIOAnimatedCoder animatedImageFrameAtIndex:] (SDImageIOAnimatedCoder.m:0)
7
GoShare2
-[SDAnimatedImage animatedImageFrameAtIndex:] (SDAnimatedImage.m:0)
8
GoShare2
__43-[SDAnimatedImagePlayer displayDidRefresh:]_block_invoke (SDAnimatedImagePlayer.m:0)
9
Foundation
0x0000000182778000 + 333168
21
libsystem_pthread.dylib
_pthread_wqthread + 14282792934007898340
#0 Thread

0
libsystem_kernel.dylib
mach_msg_trap + 8
7
UIKitCore
UIApplicationMain + 12777678613570914380
8
GoShare2
0x0000000102adc000 + 16291224644985627692
9
???
0x00000001047fc000 + 0
#1 AVAudioSession Notify Thread

0
libsystem_kernel.dylib
mach_msg_trap + 8
4
CoreFoundation
CFRunLoopRunSpecific + 10122455889026744920
5
AudioSession
0x000000018a2ce000 + 8950564510305903368
6
AudioSession
0x000000018a2ce000 + 5294009949276405092
7
libsystem_pthread.dylib
_pthread_start + 11974046664341061780
#2 com.taobao.weex.bridge

0
libsystem_kernel.dylib
mach_msg_trap + 8
5
Foundation
0x0000000182778000 + 17520070176261901652
6
GoShare2
0x0000000102adc000 + 11544222280286639604
7
Foundation
0x0000000182778000 + 428380
8
libsystem_pthread.dylib
_pthread_start + 9618201164830998676
#3 com.apple.uikit.eventfetch-thread

0
libsystem_kernel.dylib
mach_msg_trap + 8
9
libsystem_pthread.dylib
_pthread_start + 2113062886736658580
#4 JavaScriptCore bmalloc scavenger

0
libsystem_kernel.dylib
__psynch_cvwait + 8
7
libsystem_pthread.dylib
_pthread_start + 17748429395502039188
#5 WebThread

0
libsystem_kernel.dylib
mach_msg_trap + 8
6
libsystem_pthread.dylib
_pthread_start + 16945264738713993364
#6 org.hxhg.jiguang.thread

0
libsystem_kernel.dylib
mach_msg_trap + 8
6
Foundation
0x0000000182778000 + 583506422814315792
7
GoShare2
0x0000000102adc000 + 15370965848134309800
8
Foundation
0x0000000182778000 + 428380
9
libsystem_pthread.dylib
_pthread_start + 6058797801964306580
#7 io.jiguang.hxhg.cache

0
libsystem_kernel.dylib
mach_msg_trap + 8
5
CoreFoundation
CFRunLoopRun + 16893692695568973888
6
GoShare2
0x0000000102adc000 + 10837223109488634916
7
Foundation
0x0000000182778000 + 428380
8
libsystem_pthread.dylib
_pthread_start + 17013145288078000276
#8 io.jiguang.hxhg.network

0
libsystem_kernel.dylib
mach_msg_trap + 8
5
CoreFoundation
CFRunLoopRun + 14353799944685486144
6
GoShare2
0x0000000102adc000 + 1892837854530598696
7
Foundation
0x0000000182778000 + 428380
8
libsystem_pthread.dylib
_pthread_start + 17609510499379052692
#9 com.apple.CFSocket.private

0
libsystem_kernel.dylib
__select + 8
2
libsystem_pthread.dylib
_pthread_start + 16463176168934211732
#10 com.apple.NSURLConnectionLoader

0
libsystem_kernel.dylib
mach_msg_trap + 8
7
libsystem_pthread.dylib
_pthread_start + 9972795314056200340
#11 com.taobao.weex.component

0
libsystem_kernel.dylib
mach_msg_trap + 8
5
Foundation
0x0000000182778000 + 13122858194483383636
6
GoShare2
0x0000000102adc000 + 11129511783057148176
7
Foundation
0x0000000182778000 + 428380
8
libsystem_pthread.dylib
_pthread_start + 9735040218200801428
#12 Thread

0
libsystem_kernel.dylib
__workq_kernreturn + 8
1
libsystem_pthread.dylib
_pthread_wqthread + 364
#13 Thread

0
libsystem_kernel.dylib
__rename + 8
2
Foundation
0x0000000182778000 + 18311202877307344680
3
GoShare2
__cxa_throw + 12717529280217631604
4
Foundation
0x0000000182778000 + 333168
16
libsystem_pthread.dylib
_pthread_wqthread + 4551604854795010276
#14 Thread

0
libsystem_kernel.dylib
__workq_kernreturn + 8
1
libsystem_pthread.dylib
_pthread_wqthread + 364
#15 Thread

0
libsystem_kernel.dylib
__workq_kernreturn + 8
1
libsystem_pthread.dylib
_pthread_wqthread + 364
#16 Thread

0
libsystem_kernel.dylib
__workq_kernreturn + 8
1
libsystem_pthread.dylib
_pthread_wqthread + 364

**Comment by kinarobin** [comment]
```objc
Thread 44 EXC_BREAKPOINT (SIGTRAP)

0 | CoreFoundation | _CFDataGetBytes()
1 | ImageIO | _CGImageGetImageSource()
2 | ImageIO | _CGImageGetImageSource()
3 | UIKitCore | -[_UIImageCGImageContent initWithCGImage:scale:]()
4 | UIKitCore | -[UIImage initWithCGImage:scale:orientation:]()
5 | app | +[SDImageIOAnimatedCoder createFrameAtIndex:source:scale:preserveAspectRatio:thumbnailSize:options:](SDImageIOAnimatedCoder.m:293)
6 | app | -[SDImageIOAnimatedCoder animatedImageFrameAtIndex:](SDImageIOAnimatedCoder.m:0)
7 | app | -[SDAnimatedImage animatedImageFrameAtIndex:](SDAnimatedImage.m:0)
8 | app | __43-[SDAnimatedImagePlayer displayDidRefresh:]_block_invoke(SDAnimatedImagePlayer.m:0)
9 | Foundation | ___NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__()
10 | Foundation | -[NSBlockOperation main]()
11 | Foundation | ___NSOPERATION_IS_INVOKING_MAIN__()
12 | Foundation | -[NSOperation start]()
13 | Foundation | ___NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__()
14 | Foundation | ___NSOQSchedule_f()
15 | libdispatch.dylib | __dispatch_block_async_invoke2()
16 | libdispatch.dylib | __dispatch_client_callout()
17 | libdispatch.dylib | __dispatch_continuation_pop()
18 | libdispatch.dylib | __dispatch_async_redirect_invoke()
19 | libdispatch.dylib | __dispatch_root_queue_drain()
20 | libdispatch.dylib | __dispatch_worker_thread2()
21 | libsystem_pthread.dylib | __pthread_wqthread()

**Comment by tomsemi** [comment]
We also have the same problem

**Comment by jschedler-imvu** [comment]
We released our app where we replace SDAnimatedImageView and UIImageView and saw the crash disappear.  There could be two reason for this.
1.  Maybe SDAnimatedImageView is not safe for iOS 15 (threads?)
2. We were always setting known static images to SDAnimatedImageView in a location where at least 8 SDAnimatedImageView were loading at the same time.  Maybe SDAnimatedImageView + static images + iOS 15 is

**Comment by kinarobin** [comment]
@WJustin @Johnlee666 @tomsemi @jschedler-imvu Could you take a look in #3276 ?

**Comment by waydelau** [comment]
We also have the same problem



Crashed: NSOperationQueue 0x1203e7050 (QOS: UNSPECIFIED)
0  CoreFoundation                 0x18243a124 CFDataGetBytes + 156
1  ImageIO                        0x183eaf700 CGImageGetImageSource + 156
2  UIKitCore                      0x184a3dc74 -[_UIImageCGImageContent initWithCGImage:scale:] + 40
3  UIKitCore                      0x1849c4ee0 -[UIImage initWithCGImage:scale:orientation:] + 76
4  SDWebImage                     0x1070bdc2c +[SDImageIOAnimatedCoder createFrameAtIndex:source:scale:preserveAspectRatio:thumbnailSize:options:] + 270 (SDImageIOAnimatedCoder.m:270)
5  SDWebImage                     0x1070bf6d0 -[SDImageIOAnimatedCoder animatedImageFrameAtIndex:] + 653 (SDImageIOAnimatedCoder.m:653)
6  SDWebImage                     0x1070ab1dc -[SDAnimatedImage animatedImageFrameAtIndex:] + 263 (SDAnimatedImage.m:263)
7  SDWebImage                     0x1070ace68 __43-[SDAnimatedImagePlayer displayDidRefresh:]_block_invoke + 352 (SDAnimatedImagePlayer.m:352)
8  Foundation                     0x183bf9570 __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 24
9  Foundation                     0x183c0b6d8 -[NSBlockOperation main] + 104
10 Foundation                     0x183be4124 __NSOPERATION_IS_INVOKING_MAIN__ + 24
11 Foundation                     0x183bf5010 -[NSOperation start] + 804
12 Foundation                     0x183bf85b8 __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 24
13 Foundation                     0x183c06630 __NSOQSchedule_f + 184
14 libdispatch.dylib              0x182098600 _dispatch_block_async_invoke2 + 148
15 libdispatch.dylib              0x182089950 _dispatch_client_callout + 20
16 libdispatch.dylib              0x18208ce04 _dispatch_continuation_pop + 504
17 libdispatch.dylib              0x18208c4f8 _dispatch_async_redirect_invoke + 748
18 libdispatch.dylib              0x18209af48 _dispatch_root_queue_drain + 388
19 libdispatch.dylib              0x18209b768 _dispatch_worker_thread2 + 164
20 libsystem_pthread.dylib        0x1f1f4d174 _pthread_wqthread + 228
21 libsystem_pthread.dylib        0x1f1f4cf50 start_wqthread + 8

**Comment by dreampiggy** [comment]
Fixed in [v5.12.0](https://github.com/SDWebImage/SDWebImage/releases/tag/5.12.0)

**Comment by anhnguyen123** [comment]
still error at v5.12.0 @dreampiggy 

SDImageIOAnimatedCoder createFrameAtIndex:source:scale:preserveAspectRatio:thumbnailSize:options:]
![image](https://user-images.githubusercontent.com/18522542/135564992-08910c7c-07e8-4a18-8ba4-eaea024ac2cc.png)

**Comment by MainasuK** [comment]
I think that's the same issue: https://github.com/SDWebImage/SDWebImage/issues/3068

The APNG animated image cause exception sometimes. 
I make a demo app to reproduce this issue.
Repo: https://github.com/MainasuK/Issue-memmove

This issue is related to memory. I test on a real device and after three times crash the issue did not happen again. But that also happens on the simulator. After running 3min the simulator crash with the same callback stack.

![image](https://user-images.githubusercontent.com/7940186/136648217-77ff8846-3db5-440f-b47d-fae8747b748f.png)

Test on Xcode 13.0 (13A233)

**Comment by MainasuK** [comment]
> @MainasuK I can't reproduce with your demo in iPhone 12 & iOS 15 & Xcode 13.0 or simulator.

It's now always happened. I get that crash 1 in 3 runs. Could you try again? 
On my machine, it's happening 1 in 3 runs.

**Comment by dreampiggy** [comment]
为什么UIImage不能后台调用？Apple文档标记了是线程安全的，如果不是，应该Feedback给Apple。
大部分代码都依赖UIImage，ImageIO的线程安全书写的

**Comment by dreampiggy** [comment]
See: https://developer.apple.com/documentation/uikit/uiimage?language=objc

> Because image objects are immutable, you cannot change their properties after creation. Most image properties are set automatically using metadata in the accompanying image file or image data. The immutable nature of image objects also means that they are safe to create and use from any thread.

**Comment by a130785** [comment]
> @a130785 The crash stack is same as above?

是的

**Comment by tomsemi** [comment]
same issue

**Comment by penfeizhou** [comment]
### This following API is not thread safe. 

```objective-c
IMAGEIO_EXTERN CGImageRef _iio_Nullable CGImageSourceCreateImageAtIndex(CGImageSourceRef _iio_Nonnull isrc, size_t index, CFDictionaryRef _iio_Nullable options)  IMAGEIO_AVAILABLE_STARTING(10.4, 4.0);
```
https://developer.apple.com/documentation/imageio/1465011-cgimagesourcecreateimageatindex?language=objc

When the same url with animated image such as apng(frames more than 20) is  loaded into 5 or 6 SDAnimatedImageViews, app will be easily crashed.

In SDWebImage,  `SDAnimatedImagePlayer`  create animated frame serially. But  SD caches `SDAnimatedImage` in memory, so there will be 5 or 6 `SDAnimatedImagePlayer` instances call just one `SDAnimatedImage` as `animatedProvider` here to `createFrameAtIndex`, which leads to  the same `SDImageIOAnimatedCoder` instance here. Therefor `CGImageSourceCreateImageAtIndex` could be called concurrently.That leads the crash.

I suggest the situation(single `SDAnimatedImage`, multi `SDAnimatedImageView`)should be considered.

**Comment by a130785** [comment]
This is a question of multithreading, the options parameter SDWebImagePreloadAllFrames can solve

**Comment by andrey-shostik** [comment]
Does the workaround fixed the crashes? How to apply the fix?

**Comment by MainasuK** [comment]
> Does the workaround fixed the crashes? How to apply the fix?

https://github.com/MainasuK/Issue-memmove
I run my issue check demo 3 times without crash.

But I'm not sure. Because I just change my development device from Mac mini to MBP. So still needs more time to verify it.

**Comment by waydelau** [comment]
Add a thread-safe lock to fix this bug.After I tried, this bug did not appear again in my APP. Of course, I am not sure whether this modification will lose performance.

![企业微信截图_9d0cb461-3315-4fc1-b2d8-41d121c4b068](https://user-images.githubusercontent.com/3348044/142173256-8d11ed3e-48f0-4295-99ee-abd3d0ceb7bf.png)
![企业微信截图_a089a6aa-e0da-4682-99d2-72eb92fc49cc](https://user-images.githubusercontent.com/3348044/142173277-1af09385-a6e6-4e42-989d-a28cd116774c.png)

![企业微信截图_da711c17-780d-4802-9e8c-87c4108f697a](https://user-images.githubusercontent.com/3348044/142173287-5b9fcc8c-9a96-4832-9eac-4ede21a230f8.png)

**Comment by huangjianwengit** [comment]
my app have same question ,I use SDWebImage (5.12.1)

**Comment by weiningtry7** [comment]
Why this commit https://github.com/SDWebImage/SDWebImage/pull/3276 was reverted， is any  other question, i want know
the suitable solution

**Comment by kinarobin** [comment]
@weiningtry7 This fix #3276  still had crash.

**Comment by Johnlee666** [comment]
> @ACFancy 你可以帮忙提一个Radar Feedback给Apple，里面带上这个Demo，以及文档链接
请问这里苹果有回复吗，确实是苹果的bug导致的线程安全的问题？

**Comment by tomsemi** [comment]
> Add a thread-safe lock to fix this bug.After I tried, this bug did not appear again in my APP. Of course, I am not sure whether this modification will lose performance.
> 
> ![企业微信截图_9d0cb461-3315-4fc1-b2d8-41d121c4b068](https://user-images.githubusercontent.com/3348044/142173256-8d11ed3e-48f0-4295-99ee-abd3d0ceb7bf.png) ![企业微信截图_a089a6aa-e0da-4682-99d2-72eb92fc49cc](https://user-images.githubusercontent.com/3348044/142173277-1af09385-a6e6-4e42-989d-a28cd116774c.png)
> 
> ![企业微信截图_da711c17-780d-4802-9e8c-87c4108f697a](https://user-images.githubusercontent.com/3348044/142173287-5b9fcc8c-9a96-4832-9eac-4ede21a230f8.png)

Has this fix been verified on a large scale?

**Comment by tomsemi** [comment]
#3304

**Comment by tomsemi** [comment]
> Add a thread-safe lock to fix this bug.After I tried, this bug did not appear again in my APP. Of course, I am not sure whether this modification will lose performance.
> 
> ![企业微信截图_9d0cb461-3315-4fc1-b2d8-41d121c4b068](https://user-images.githubusercontent.com/3348044/142173256-8d11ed3e-48f0-4295-99ee-abd3d0ceb7bf.png) ![企业微信截图_a089a6aa-e0da-4682-99d2-72eb92fc49cc](https://user-images.githubusercontent.com/3348044/142173277-1af09385-a6e6-4e42-989d-a28cd116774c.png)
> 
> ![企业微信截图_da711c17-780d-4802-9e8c-87c4108f697a](https://user-images.githubusercontent.com/3348044/142173287-5b9fcc8c-9a96-4832-9eac-4ede21a230f8.png)

Has this fix been proved on a large scale?

**Comment by JivanHuang** [comment]
> > Add a thread-safe lock to fix this bug.After I tried, this bug did not appear again in my APP. Of course, I am not sure whether this modification will lose performance.
> > ![企业微信截图_9d0cb461-3315-4fc1-b2d8-41d121c4b068](https://user-images.githubusercontent.com/3348044/142173256-8d11ed3e-48f0-4295-99ee-abd3d0ceb7bf.png) ![企业微信截图_a089a6aa-e0da-4682-99d2-72eb92fc49cc](https://user-images.githubusercontent.com/3348044/142173277-1af09385-a6e6-4e42-989d-a28cd116774c.png)
> > ![企业微信截图_da711c17-780d-4802-9e8c-87c4108f697a](https://user-images.githubusercontent.com/3348044/142173287-5b9fcc8c-9a96-4832-9eac-4ede21a230f8.png)
> 
> Has this fix been proved on a large scale?

My app didn't crash again after I did that

**Comment by waydelau** [comment]
> > Add a thread-safe lock to fix this bug.After I tried, this bug did not appear again in my APP. Of course, I am not sure whether this modification will lose performance.
> > ![企业微信截图_9d0cb461-3315-4fc1-b2d8-41d121c4b068](https://user-images.githubusercontent.com/3348044/142173256-8d11ed3e-48f0-4295-99ee-abd3d0ceb7bf.png) ![企业微信截图_a089a6aa-e0da-4682-99d2-72eb92fc49cc](https://user-images.githubusercontent.com/3348044/142173277-1af09385-a6e6-4e42-989d-a28cd116774c.png)
> > ![企业微信截图_da711c17-780d-4802-9e8c-87c4108f697a](https://user-images.githubusercontent.com/3348044/142173287-5b9fcc8c-9a96-4832-9eac-4ede21a230f8.png)
> 
> Has this fix been proved on a large scale?

Yes, this crash was resolved after the new version was released.

**Comment by CaiChenghan** [comment]
This problem still exist and my app crashed，**SDWebImage** version: **5.12.4**

**Comment by alex-nikolaichuk** [comment]
This problem still exist and my app crashed，SDWebImage version: 5.12.4

**Comment by sikandarali307** [comment]
this problem still exit bo[](url
<img width="1028" alt="Screenshot 2022-04-26 at 1 11 43 PM" src="https://user-images.githubusercontent.com/35461226/165253601-0249a997-7fac-47be-8cb8-14fb4e8e59ff.png">
)

**Comment by usercai** [comment]
This problem still exist and my app crashed
![image](https://user-images.githubusercontent.com/16284097/165454916-a7effd54-352b-45f3-afb4-dc87c979b622.png)

**Comment by yunjinghui123** [comment]
This problem still exist and my app crashed. 5.12.5
<img width="1165" alt="image" src="https://user-images.githubusercontent.com/9649401/167249703-8c71fd48-1fc8-4cb7-9dae-d482c5a0d7d9.png">
Just don't show the picture. Don't crash。

**Comment by lianleven** [comment]
Locking does work, I did not have this crash on the online project.

**Comment by yunjinghui123** [comment]
Locking will cause the page to become suddenly stuck?

**Comment by yunjinghui123** [comment]
Add a thread-safe lock to fix this bug.My app didn't crash again after I did that

**Comment by ChengQuanYe** [comment]
So has this issue been resolved?

**Comment by Deco354** [comment]
> So has this issue been resolved?

In the main repo no.

If you use one of the many forks implementing this change it appears it is. I've opened a [PR](https://github.com/SDWebImage/SDWebImage/pull/3384) to merge this in to the main Repo so we don't have to rely on forks.

**Comment by dreampiggy** [comment]
Honestly saying. Let me tell the reason:

### Background

There used to be a C++ deconstructor multi-thread crash (where only happended during you process finished, so no visible to end-user), which fixed by https://github.com/SDWebImage/SDWebImage/pull/3276. And available in v5.12.0 version

However, @kinarobin found even in v5.12.0 version, that crash still exists, so he revert again in v5.12.1 version via https://github.com/SDWebImage/SDWebImage/pull/3281

### Current Status

This PR seems talke another issue, which shows up the `CGImageSource` and `CGImageDestination` API is **no longer thread safe**. And there are some other issues point it out: https://github.com/SDWebImage/SDWebImage/issues/3099

This is an unexpected change behavior for me, because ImageIO is used to thread-safe, so we design many multi-thread decoder to make it fast.

Until Apple release iOS 16 to see whether this thread-unsafe is indeed (or just iOS 15 bug), I think we can not simply add some lock between CGImageSource call. Which will slow down other platform user like iOS 9~iOS 14

**Comment by dreampiggy** [comment]
Or another possible reason I guess, for iOS 15

> From iOS 15, CGImageSource created CGImageRef will retain the `CGImageRef`

This may cause issue in multi-thread environment. For our SDAnimatedImageView example

1. Display Link refresh
2. CGImageSourceRef create CGImageRef in global queue 1
3. UIImageView retain UIImage, which retain CGImageRef, which retain the CGImageSourceRef from main queue
4. CGImageSourceRef try to get data buffer, which may be lazy and create on access. The queue 1 and main queue access cause the crash

**Comment by Deco354** [comment]
> But this need a real App testing for large dataset. I can provide the code, but need someone here to test it on their App or simulate the multi-thread cases

We're willing to do this with the Tumblr app. Send us a branch or feel free to make one on our fork and we'll put it in our next build.

**Comment by Deco354** [comment]
> Until Apple release iOS 16 to see whether this thread-unsafe is indeed (or just iOS 15 bug)

Before applying the fix we had 1% of our user base on iOS 16 and they represented 1% of our SDImageIOAnimatedCoder crashes so at least for now this is not fixed on iOS 16.

> I think we can not simply add some lock between CGImageSource call. Which will slow down other platform user like iOS 9~iOS 14

I could modify [this PR](https://github.com/SDWebImage/SDWebImage/pull/3384) so the lock solution is only applied to iOS 15+?

**Comment by dreampiggy** [comment]
@Deco354 @ChengQuanYe @yunjinghui123 @lianleven @@sikandarali307 @alex-nikolaichuk @alex-nikolaichuk 

For people who still facing this issue on iOS 15+, please using code here to use branch dependency or copy to your codebase: https://github.com/SDWebImage/SDWebImage/pull/3387

Please report me the feedback (whether new version crash ratio get down or not)

**Comment by Deco354** [comment]
Thanks @dreampiggy. This will ship in Tumblr's beta tomorrow and production this time next week. We'll keep you posted on the results.

**Comment by Deco354** [comment]
> 
> I could modify [this PR](https://github.com/SDWebImage/SDWebImage/pull/3384) so the lock solution is only applied to iOS 15+?

I've done this as a backup for the time being. Hopefully the [CGImageSource retain solution](https://github.com/SDWebImage/SDWebImage/pull/3387) works. We should have data on this soon.

**Comment by lbo7544** [comment]
> @Deco354 @ChengQuanYe @yunjinghui123 @lianleven @@sikandarali307 @alex-nikolaichuk @alex-nikolaichuk
> 
> For people who still facing this issue on iOS 15+, please using code here to use branch dependency or copy to your codebase: #3387
> 
> Please report me the feedback (whether new version crash ratio get down or not)

We used the modified code in the new version, but the same crash still occurred

**Comment by dreampiggy** [comment]
> > I could modify [this PR](https://github.com/SDWebImage/SDWebImage/pull/3384) so the lock solution is only applied to iOS 15+?
> 
> I've done this as a backup for the time being. Hopefully the [CGImageSource retain solution](https://github.com/SDWebImage/SDWebImage/pull/3387) works. We should have data on this soon.

Any news about your testing ?

**Comment by Deco354** [comment]
@dreampiggy We'll have data by Wednesday I expect. The beta didn't give us any crashes but it's just our QA team on that unfortunately. Production released the hotfix a few hours a go so we'll soon see the crashes if there are any.

**Comment by Deco354** [comment]
The crashes are back for us too I'm afraid.

<details>
<summary>Crash Log</summary>

```
Crashed: NSOperationQueue 0x15cedaad0 (QOS: UNSPECIFIED)
0  CoreFoundation                 0xbc1c4 CFDataGetBytes + 156
1  ImageIO                        0x1adc CGImageGetImageSource + 156
2  UIKitCore                      0x215284 -[_UIImageCGImageContent initWithCGImage:scale:] + 40
3  UIKitCore                      0x19c8d8 -[UIImage initWithCGImage:scale:orientation:] + 76
4  Tumblr                         0x1547bbc +[SDImageIOAnimatedCoder createFrameAtIndex:source:scale:preserveAspectRatio:thumbnailSize:options:] + 247 (SDImageIOAnimatedCoder.m:247)
5  Tumblr                         0x1549544 -[SDImageIOAnimatedCoder animatedImageFrameAtIndex:] + 647 (SDImageIOAnimatedCoder.m:647)
6  Tumblr                         0x1534fd4 -[SDAnimatedImage animatedImageFrameAtIndex:] + 263 (SDAnimatedImage.m:263)
7  Tumblr                         0x1536ba4 __43-[SDAnimatedImagePlayer displayDidRefresh:]_block_invoke + 352 (SDAnimatedImagePlayer.m:352)
8  Foundation                     0x50c64 __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 24
9  Foundation                     0x62df8 -[NSBlockOperation main] + 104
10 Foundation                     0x3b784 __NSOPERATION_IS_INVOKING_MAIN__ + 24
11 Foundation                     0x4c700 -[NSOperation start] + 804
12 Foundation                     0x4fcac __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 24
13 Foundation                     0x5dd50 __NSOQSchedule_f + 184
14 libdispatch.dylib              0x13454 _dispatch_block_async_invoke2 + 148
15 libdispatch.dylib              0x4670 _dispatch_client_callout + 20
16 libdispatch.dylib              0x7b44 _dispatch_continuation_pop + 504
17 libdispatch.dylib              0x71a0 _dispatch_async_redirect_invoke + 596
18 libdispatch.dylib              0x15de0 _dispatch_root_queue_drain + 396
19 libdispatch.dylib              0x16608 _dispatch_worker_thread2 + 164
20 libsystem_pthread.dylib        0x10b8 _pthread_wqthread + 228
21 libsystem_pthread.dylib        0xe94 start_wqthread + 8
```
</details>

**Comment by dreampiggy** [comment]
> createFrameAtIndex

OK. Now let me try another way. Update that #3276

In `createFrameAtIndex:`, I'll call `=`[SDImageCoderHelper CGImageCreateScaled:]` directlly before init UIImage.

This can ensure the CGImage does not bind the CGImageSource.

**Comment by Deco354** [comment]
@dreampiggy Thanks for updating the fix. 

It will enter Tumblr's beta next week and production the week after.
We'll have our first day of production data approximately around 15th September and should have a weeks worth around the 21st.

I'll post updates if our QA team or performance metrics flag up the beta or around the 15th when we've data on if the crash has disappeared.

**Comment by rezq2009** [comment]
Hello there

Use the latest version 5.13.2 - 5.13 Patch and i face the same issue and app crashed

Details in the picture

<img width="1914" alt="Screen Shot 2022-09-14 at 4 23 27 PM" src="https://user-images.githubusercontent.com/102135523/190167786-f3689bc7-9976-47fc-a1ad-152c31de50f4.png">

**Comment by dreampiggy** [comment]
> Hello there
> 
> Use the latest version 5.13.2 - 5.13 Patch and i face the same issue and app crashed
> 
> Details in the picture
> 
> <img alt="Screen Shot 2022-09-14 at 4 23 27 PM" width="1914" src="https://user-images.githubusercontent.com/102135523/190167786-f3689bc7-9976-47fc-a1ad-152c31de50f4.png">

Please use the branch in #3387 to see whether this been fixed.

**Comment by rezq2009** [comment]
Hello @dreampiggy  

Thanks for your attention
i have succeeded in build use latest version 5.13.2  && branch in https://github.com/SDWebImage/SDWebImage/pull/3387

Now I'm doing deep Test , I'll get back to you here with the results

Thanks

**Comment by rezq2009** [comment]
Hi @dreampiggy  

So far I tested branch https://github.com/SDWebImage/SDWebImage/pull/3387 for a whole day and the error didn't appear again, and the app do not crashed never .. i have hard test with more Gif images in same screen and do not crashed , it seems that fixed !

However, I will continue the test for two more days and will post another comment here later

Thanks

**Comment by Deco354** [comment]
We've 48 hours worth of data for the new fix on PROD for Tumblr now and no crashes are appearing with the new fix yet. It's looking good so far 🤞. There are delays in our crash reporting and the uptake to the app upgrade is immediate so I'll report back when we've more data.

Our gif download/render metrics appear fine too although we may need more time for the stats on that to converge.

**Comment by rezq2009** [comment]
Hello

@Deco354  @dreampiggy 

My Final Review after tested 3 Days 

branch https://github.com/SDWebImage/SDWebImage/pull/3387 

The previous error i mentioned (https://github.com/SDWebImage/SDWebImage/issues/3273#issuecomment-1246771506) has been Fixed !!
No Any Crash !!

Thanks ..

**Comment by dreampiggy** [comment]
Seems it's time for that #3387 to be merged and release new patch version.

**Comment by rezq2009** [comment]
> Seems it's time for that #3387 to be merged and release new patch version.

This must be done quickly, as this repair branch https://github.com/SDWebImage/SDWebImage/pull/3387 is very important and considered a serious issue
I even noticed an improvement in performance as well as the end of a crash issue

**Comment by dreampiggy** [comment]
[5.13.3](https://github.com/SDWebImage/SDWebImage/releases/tag/5.13.3) released.

Please use the version for testing, provide more feedback if you face the similiar crash stack trace.

**Comment by rezq2009** [comment]
Hello

im trying to use [5.13.3](https://github.com/SDWebImage/SDWebImage/releases/tag/5.13.3) released.

But i got an error :


[!] CocoaPods could not find compatible versions for pod "SDWebImage":

      SDWebImage (~> 5.13.3)

None of your spec sources contain a spec satisfying the dependency: `SDWebImage (~> 5.13.3)`.

You have either:
 * out-of-date source repos which you can update with `pod repo update` or with `pod install --repo-update`.
 * mistyped the name or version.
 * not added the source repo that hosts the Podspec to your Podfile.


*thinking it may take some time to become available ..
any recommendations ?

**Comment by hstdt** [comment]
@dreampiggy It seems fixed this problem.

But another crash appears on iOS 13. `dyld: lazy symbol binding failed: Symbol not found:`

<img width="1326" alt="CleanShot 2022-09-20 at 19 33 37" src="https://user-images.githubusercontent.com/10215098/191247243-f9cccdbc-1cda-4025-b3b1-ef7bf985a6c6.png">

PS: the `Backport` structure is conflict with https://github.com/shaps80/SwiftUIBackports

**Comment by dreampiggy** [comment]
That SwiftUI related crash, please fire a new issue in : https://github.com/SDWebImage/SDWebImageSwiftUI/issues. Which seems introduced in SDWebImageSwiftUI v2.1.0

**Comment by Deco354** [comment]
Thanks for all your hard work on this @dreampiggy.
Would it be possible to use #3389 to resolve this until a more elegant solution can be found? I've restricted it to only be applied for iOS 15 and later.

**Comment by dreampiggy** [comment]
Your screenshot is another issue.

This PR and issue talked about this stack trace crash

```
0 | CoreFoundation | _CFDataGetBytes()
1 | ImageIO | _CGImageGetImageSource()
```

**Comment by Deco354** [comment]
> Your screenshot is another issue.
> 
> This PR and issue talked about this stack trace crash
> 
> ```
> 0 | CoreFoundation | _CFDataGetBytes()
> 1 | ImageIO | _CGImageGetImageSource()
> ```

@dreampiggy I believe this is related. We didn't get this crash until introducing #3387, the frequency is roughly equivalent and SDImageIOAnimatedCoder: 256 appears to be the same line as SDImageIOAnimatedCoder: 247 prior to the 5.13.3 patch.

Perhaps I'm missing something can you explain how this is different? I've searched our own SDImageIOAnimatedCoder crashes and I can't find the trace you've posted.

**Comment by dreampiggy** [comment]
> Your screenshot is another issue.
> 
> This PR and issue talked about this stack trace crash
> 
> ```
> 0 | CoreFoundation | _CFDataGetBytes()
> 1 | ImageIO | _CGImageGetImageSource()
> ```

This screenshot, it's the same, UIKit call `_CGImageGetImageSource` when submitting image content to CALayerRenderServer

https://user-images.githubusercontent.com/29303030/137888240-ca0c7f16-ca25-48bd-91eb-709a7efe8a2c.png

**Comment by Deco354** [comment]
You're right the `SDImageCoderHelper: 315` containing an `SDImageIOAnimatedCoder` appears to be an older crash that has simply changed signature on Crashlytics for us.
We'll have thousands less crashes per month due to this fix. Thanks @dreampiggy 🍾

**Comment by rezq2009** [comment]
> [5.13.3](https://github.com/SDWebImage/SDWebImage/releases/tag/5.13.3) released.
> 
> Please use the version for testing, provide more feedback if you face the similiar crash stack trace.

UPDATE : Since it was released and I'm testing , so far no any crash has happened !! 

I am very surprised by the responses of the guys above, how this crash happens with them, I do not know

Thanks

**Comment by rezq2009** [comment]
Hello @dreampiggy 

Got it , waiting for 5.13.4 

Thank you very much, your efforts are greatly appreciated !

**Comment by dreampiggy** [comment]
[5.13.4](https://github.com/SDWebImage/SDWebImage/releases/tag/5.13.4) Released

**Comment by Deco354** [comment]
> Crash signature after #3387 ![image](https://user-images.githubusercontent.com/18557720/191292861-fa27c87a-ad44-4776-a857-8ba2f0ff7c84.png) ![Screen Shot 2022-09-20 at 16 13 31](https://user-images.githubusercontent.com/18557720/191297015-e0994962-8374-4cd5-b8f4-b4aa83b8f731.png)
> 
> Crash log

On the off-chance someone else gets this crash I mistook for the crash this issue references 5.13.4 fixed it for us.

**Comment by szhang-code** [comment]
> 

@dreampiggy 
hi, Thanks for the analysis and efforts. According to feedbacks, the fix seems to have been proved to work. But I still have some questions and can't not figure it out myself. Could you please explain a bit more? 
1. which one is not thread-safe, CGImageSourceRef or CGImageRef? why does CGImageSourceRef's accessing data buffer from multiple threads at same time cause crashes? Is it Apple's bug? does it only happen on iOS 15+
2.does +[SDImageCoderHelper CGImageCreateDecoded:] need access to CGImageSourceRef  in Queue 1, which is same with creating UIImage from CGImageRef using -[UIImage imageWithCGImage:]?
3.As @penfeizhou pointed out above, Will multiple SDAnimatedImagePlayer playing at same time with a single animated image cause problems?

**Comment by dreampiggy** [comment]
Hi. Anyone.

In 5.13.4, I use the`Force Decode` (which means, create and allocate a new CGImageContext and draw the image), to strip down the CGImageSource to avoid the thread-safe issue.

I found there are another way to solve this, by using the `CGImageCreate` with all information from CGImage. So I will use the new solution instead on SDWebImage 5.14.0

Which may reduce the RAM usage if you don't want `Force Decode`, by using `SDWebImageAvoidDecodeImage` option (or directly call SDAnimatedImage with `SDImageCoderDecodeAvoidLazyDecoding = NO`)

**Comment by rezq2009** [comment]
Hi @dreampiggy  

I will test it as soon as it is released and let you know the results as soon as possible

thanks for your efforts

**Comment by dreampiggy** [comment]
[5.14.0](https://github.com/SDWebImage/SDWebImage/releases/tag/5.14.0) released.

If people who used to use lazy-decoding on SDAnimatedImageView (This is the behavior from 5.12.0 <-> 5.13.2, but changed after 5.13.2), who can opt-in by passing `.decodeUseLazyDecoding = YES` to context option `.imageDecodeOptions`.

Though my personal thinking that Animated Image like GIF, should avoid lazy-decoding because the rendering speed is fast, lazy-decoding will block main queue until the decoding finished. So the default behavior is still `non-lazy-decoding` for animated image.

But `non-lazy-decoding` may increse RAM usage for large images. Thanksfully that SDAnimatedImageView has one `maxBufferSize` to limit the size, so it's OK. But different people may face different use case, if you think more about RAM usage, use lazy-decoding is OK

**Comment by dreampiggy** [comment]
I found that #3425 break the non-animated image JPEG && HEVC's optimization, which used to save 50% RAM usage by using the YUV422 sample instead of RGBA8888

So, instead, I opt-in my #3425, now the JPEG && HEVC format does not use that `strip CGImageSource from CGImage` solution and keep going. I only strip the format like GIF/APNG

See more (Chinese) about how Apple do this optimization without providing any public API : https://dreampiggy.com/2022/11/07/%E6%8E%A2%E7%A7%98%E8%B6%8A%E6%9D%A5%E8%B6%8A%E5%A4%8D%E6%9D%82%E7%9A%84ImageIO%E6%A1%86%E6%9E%B6/
